### PR TITLE
Add a way to disable checksum mode for incompatible s3 providers

### LINKED ISSUE #26394: Video / Audio uploads fails with Cloudflare R2
### Steps to reproduce the problem

1. Upload a video file

### Expected behaviour

Video upload works

### Actual behaviour

Video upload doesn't work

### Detailed description

Sidekiq fails to fetch the original file uploaded to s3 : 
```
I, [2023-08-08T16:34:56.401628 #7]  INFO -- : [paperclip] copying media_attachments/files/110/854/961/122/467/275/original/9546231f02c0d6d5.mp4 to local file /tmp/eccccece145f3f7b1fe05ca4a9a938fe20230808-7-30jgaa.mp4
D, [2023-08-08T16:34:56.564749 #7] DEBUG -- : [httplog] HEAD http://x.r2.cloudflarestorage.com:443/rivals-space-media/media_attachments/files/110/854/961/122/467/275/original/9546231f02c0d6d5.mp4?partNumber=1 completed with status code 501 in 0.031263 seconds
Aws::S3::Errors::Http501Error - cannot copy media_attachments/files/110/854/961/122/467/275/original/9546231f02c0d6d5.mp4 to local file /tmp/eccccece145f3f7b1fe05ca4a9a938fe20230808-7-30jgaa.mp4
D, [2023-08-08T16:34:56.723062 #7] DEBUG -- : [httplog] HEAD http://x.r2.cloudflarestorage.com:443/rivals-space-media/media_attachments/files/110/854/961/122/467/275/original/9546231f02c0d6d5.mp4 completed with status code 200 in 0.156776 seconds
2023-08-08T16:34:56.794Z pid=7 tid=3jur class=PostProcessMediaWorker jid=6fc9822e2d8c700282b9f76c ERROR: Processing media attachment 110854961122467275 failed with Error while transcoding /tmp/eccccece145f3f7b1fe05ca4a9a938fe20230808-7-30jgaa.mp4: unsupported file
```

env vars are 
```
S3_REGION=auto
S3_ENDPOINT=https://redacted.r2.cloudflarestorage.com
S3_ENABLED=true
S3_HOSTNAME=cdn.rivals.space
S3_PROTOCOL=https
S3_PERMISSION=
S3_ALIAS_HOST=cdn.rivals.space
S3_BUCKET=redacted
AWS_SECRET_ACCESS_KEY=redacted
AWS_ACCESS_KEY_ID=redacted
```

### Mastodon instance

rivals.space

### Mastodon version

v4.1.6+nightly-2023-08-08

### Technical details

Using 2023-08-08 nightly docker image

### PR DESCRIPTION
**Pull request by emilweth**
Fix #26394

### CODE REVIEW

File: config/initializers/paperclip.rb [hunk 1]
```diff
@@ -88,6 +88,16 @@
 
   Paperclip::Attachment.default_options[:s3_headers]['X-Amz-Storage-Class'] = ENV['S3_STORAGE_CLASS'] if ENV.has_key?('S3_STORAGE_CLASS')
 
+  # Some S3 providers are not compatible with the checksum mode
+  # See https://github.com/mastodon/mastodon/issues/26394
+  if ENV['S3_DISABLE_CHECKSUM_MODE'] == 'true'
+    class Aws::S3::FileDownloader
+      def validate!
+        @params.delete(:checksum_mode)
+      end
```
Comment (ClearlyClaire) [inline config/initializers/paperclip.rb:97]: This is fine, if very hackish. If going with this, I would still call upstream's `validate!` method:
```suggestion
      def validate!
        @params.delete(:checksum_mode)
        super
      end
```

Ideally, I'd prefer if `aws-s3-sdk` provided a way to disable use of that header, though.

Comment (emilweth) [reply config/initializers/paperclip.rb]: I can open an issue on their side but I have little hope

**Comment by emilweth** [comment]
Reverting the last commit since it's not working
```
WARN: NoMethodError: super: no superclass method `validate!' for #<Aws::S3::FileDownloader:0x00007f92dba999d8 @client=#<Aws::S3::Client>, @path="/tmp/544de6d6de1ef55a0136b23b7dfd400a20230809-7-j6tnxa.mp4", @mode="auto", @thread_count=10, @chunk_size=nil, @params={:bucket=>"rivals-space-media", :key=>"media_attachments/files/110/859/007/017/229/237/original/c1631779c3d45547.mp4"}, @on_checksum_validated=nil>
```

**Comment by mjankowski** [comment]
Instead of the monkey patch, can we pass in a `checksum_mode` value of `DISABLED` (or any string that is not `ENABLED`) to the paperclip/s3 config?

It looks like what's getting triggered here is that the aws FileDownloader did not previously include this option in its `params`, but after the change its both a) including the value, b) defaulting it to `ENABLED` (which passes through to the client being built, and the API, and turns on this check as a side effect).

https://github.com/aws/aws-sdk-ruby/pull/2886/files#diff-303d2330df58df308aac2bc502ca292378fb9fcea2e6f3779f5194c25c10d0fcR34

I'm not sure if this was an intentional change there, or unintentional side effect.

**Comment by emilweth** [comment]
Last modification is working as intended, media uploads works correctly now when setting `S3_DISABLE_CHECKSUM_MODE` to true

**Comment by mjankowski** [comment]
> > I'm not sure if this was an intentional change there, or unintentional side effect.
> 
> That causes R2 to fail with: `Header 'x-amz-checksum-mode' with value 'DISABLED' not implemented (Aws::S3::Errors::NotImplemented)`

Ah, yuck.

Apparently `ENABLED` is the only allowed value of that setting at the API level - https://docs.aws.amazon.com/cli/latest/reference/s3api/head-object.html#options ... and its only the absence of it (which https://github.com/aws/aws-sdk-ruby/pull/2886/files#diff-303d2330df58df308aac2bc502ca292378fb9fcea2e6f3779f5194c25c10d0fcR34 changes) that skips the check.

I feel like that has to be an unintentional side effect of their verifying methods PR ... hopefully they respond to your issue.

**Comment by emilweth** [comment]
Closed in favor of #26435

**Pull request closed without merging**